src/System.Private.CoreLib/shared/System/Text/Rune.cs

```diff
@@ -0,0 +1,724 @@
+        /// </summary>
+        /// <remarks>
+        /// The return value will be 1 or 2.
+        /// </remarks>
+        public int Utf16SequenceLength => UnicodeUtility.GetUtf16SequenceLength(ValueUnsigned);
+
+        /// <summary>
+        /// Returns the length in code units (<see cref="Utf8Char"/>) of the
+        /// UTF-8 sequence required to represent this scalar value.
+        /// </summary>
+        /// <remarks>
+        /// The return value will be 1 through 4, inclusive.
+        /// </remarks>
+        public int Utf8SequenceLength => UnicodeUtility.GetUtf8SequenceLength(ValueUnsigned);
+
+        /// <summary>
+        /// Returns the Unicode scalar value as an integer.
+        /// </summary>
+        public int Value => (int)ValueUnsigned;
+
+        internal uint ValueUnsigned { get; }
+
+        private static Rune ChangeCase(Rune rune, CultureInfo culture, bool toUpper)
+        {
+            if (culture == null)
+            {
+                throw new ArgumentNullException(nameof(culture));
+            }
+
+            var textInfo = culture.TextInfo;
+
+            Span<char> original = stackalloc char[2]; // worst case scenario = 2 code units (for a surrogate pair)
+            Span<char> modified = stackalloc char[2]; // case change should preserve UTF-16 code unit count
+
+            int charCount = rune.EncodeToUtf16(original);
+            original = original.Slice(0, charCount);
+            modified = modified.Slice(0, charCount);
+
+            if (toUpper)
+            {
+                textInfo.ChangeCaseToUpper(original, modified);
+            }
+            else
+            {
+                textInfo.ChangeCaseToLower(original, modified);
+            }
+
+            // We use simple case folding rules, which disallows moving between the BMP and supplementary
+            // planes when performing a case conversion. The helper methods which reconstruct a Rune
+            // contain debug asserts for this condition.
+
+            if (rune.IsBmp)
+            {
+                return UnsafeCreate(modified[0]);
+            }
+            else
+            {
+                return UnsafeCreate(UnicodeUtility.GetScalarFromUtf16SurrogatePair(modified[0], modified[1]));
+            }
+        }
+
+        public int CompareTo(Rune other) => this.ValueUnsigned.CompareTo(other.ValueUnsigned);
+
+        // returns the number of chars written
+        private int EncodeToUtf16(Span<char> destination)
+        {
+            Debug.Assert(destination.Length >= Utf16SequenceLength, "Caller should've provided a large enough buffer.");
+            bool success = TryEncode(destination, out int charsWritten);
+            Debug.Assert(success, "TryEncode should never fail given a large enough buffer.");
+            return charsWritten;
+        }
+
+        public override bool Equals(object obj) => (obj is Rune other) && this.Equals(other);
```

Comment:
>(obj is Rune other) [](start = 51, length = 19)

I learned from Stephen (obj is Rune other) produce more complicated IL than if you manually doing the cast. we don't know the perf impact though

Reply:
It will definitely produce less optimized IL if the type `T` on the right side of the _is_ keyword is a reference type. I'm not sure what the behavior is since this is a value type. To be fair I didn't look at this too much since Equals(object) isn't really a highly optimized code path and I was going for readability more than anything else.